### PR TITLE
various improvements to subtyping

### PR DESCRIPTION
**Pull request by vtjnash**
Improves some fast paths, and fixes some intersection bugs (by visual comparison against the more-tested obvious_subtype, and by reading a comment that explains why the code it describes is not quite valid, haha)

### CODE REVIEW

File: src/subtype.c [hunk 1]
```diff
@@ -304,11 +304,13 @@ static int obviously_disjoint(jl_value_t *a, jl_value_t *b, int specificity)
         return 0;
     if (specificity && a == (jl_value_t*)jl_typeofbottom_type)
         return 0;
-    if (jl_is_concrete_type(a) && jl_is_concrete_type(b) &&
-        // TODO: remove these 2 lines if and when Tuple{Union{}} === Union{}
-        (((jl_datatype_t*)a)->name != jl_tuple_typename ||
-         ((jl_datatype_t*)b)->name != jl_tuple_typename))
-        return 1;
+    // TODO: this would be a nice fast-path to have, unfortuanately,
+    //       datatype allocation fails to correctly cons them
```
Comment (JeffBezanson) [inline src/subtype.c]:
```suggestion
    //       datatype allocation fails to correctly hash-cons them
```

File: src/subtype.c [hunk 2]
```diff
@@ -355,25 +357,24 @@ static int obviously_disjoint(jl_value_t *a, jl_value_t *b, int specificity)
...
         size_t i;
-        for(i=0; i < np; i++) {
-            jl_value_t *ai = jl_tparam(ad,i);
-            jl_value_t *bi = jl_tparam(bd,i);
+        for (i = 0; i < np; i++) {
+            jl_value_t *ai = jl_tparam(ad, i);
+            jl_value_t *bi = jl_tparam(bd, i);
             if (jl_is_typevar(ai) || jl_is_typevar(bi))
-                continue;
+                continue; // Union{} is always in this intersection
```
Comment (JeffBezanson) [inline src/subtype.c]: I don't think this is true, e.g. if there is a non-`Union{}` lower bound. We just skip these because vars are difficult to handle.

Comment (vtjnash) [reply src/subtype.c]:
```suggestion
                continue; // it's possible that Union{} is in this intersection
```

File: src/subtype.c [hunk 3]
```diff
@@ -3226,6 +3254,35 @@ jl_value_t *switch_union_tuple(jl_value_t *a, jl_value_t *b)
     return ans;
 }
 
+int does_subtypes_concrete(jl_value_t *a)
```
Comment (JeffBezanson) [inline src/subtype.c]:
```suggestion
// `a` might have a non-empty intersection with some concrete type b even if !(a<:b) and !(b<:a)
// For example a=`Tuple{Type{<:Vector}}` and b=`Tuple{DataType}`
int might_intersect_concrete(jl_value_t *a)
```

**Pull request merged**